### PR TITLE
`General`: Limit scope of active students line charts by course start and end date

### PR DESCRIPTION
**Pull request by willergeorg**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server..
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

We currently provide two line charts displaying the amount of active students per week. Both do not take the course start and end date into account. This has the effect, that for example weeks before the official start of the course are displayed or the active students are shown up to the current week even though the course ended weeks ago. This is changed by this PR. 
### Description

I changed the chart implementation so they are not rendered if the start date is not passed yet and only contain data until the course end date if it is passed.
Therefore I also adapted the server implementation. In case that the end date of a course is already passed, the server only sends the data collected until the end date to the client.
The data sent is also adapted if the visible time span differs from the normal one, e.g. the start date was 1 week ago or the end course life time is shorter than the chart time span.
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with an active students distribution

1. Log in to Artemis
2. Navigate to Course Administration
3. Inspect the active students chart for the course. Make sure the values are displayed appropriately. Do the same for the active students chart on course detail page.
4. Edit the course and set the end date e.g. 2 weeks before the current date.
5. Check both charts again. The data should switch accordingly and the data labels should reflect the fact, that old data is displayed (x weeks ago as most right data label on course management and the corresponding ISO week on course detail).
6. Edit the course again. Delete the end date and set the start date on some point in future
7. Check the charts. No charts should be displayed, instead a lettering explains that the course did not start yet.
8. Edit the course again. Set the start date on some point later than 3 weeks before the current date.
9. Check both charts. Now the charts should start from the week of the dedicated start date.
10. Edit the course again. Set the start date more than 16 weeks before the current date. Check the charts. The chart on course management should not be affected by it. The chart on course detail page should only allow going back in time until the start date (by clicking the left arrow)
11. Edit the course again. Set the start date of the course e.g. 2 weeks ago and the end date of the course e.g. 1 week ago. Return to the charts and make sure they change appropriately.
12. Go wild with different combinations of start and end date and make sure no unexpected behavior occurs.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| File name | Branch coverage (%) | Line coverage (%) |
| -- | -- | -- |
| `CourseRescource.java` | 73 | 93 |
| `CourseService.java` | 75 | 90 |
| `course-detail-line-chart.component.ts` | 79.16 | 94.73 |
| `course-management-overview-statistics.component.ts` | 91.66 | 91.48 |

### Example screenshot how the course management chart should look after step 4:
image
For course detail (current date is in CW 10):
image

### Example after step 6 (similar for course detail):
image

### Example after step 8:
image
For course detail:
image

<details>
     <summary>Example for course detail after step 10:</summary>

https://user-images.githubusercontent.com/80622272/157844612-a29f249f-b1c3-43f1-bb44-21117589f43c.mp4

</details>

### Example for course management after step 11:
image
For course detail:
image

### CODE REVIEW

**Comment by canberkanar** [comment]
For testing step 4;
I saw that the chart on course management behaved accordingly. However, the chart within course details is stalled.
I did the testing on my local setup, with a course that previously existed. First I uploaded a file as a student to trigger an activity and confirmed it on the tables. Then I set the course end date two weeks prior to today.
image
image

**Comment by canberkanar** [comment]
Another problematic setup;
image
image
image

**Review by canberkanar — DISMISSED** [review]
Tested locally and on ts1. On local setup, the chart on course details page was loading infinitely but everything was as expected on the test server.

File: src/main/java/de/tum/in/www1/artemis/service/CourseService.java [hunk 1]
```diff
@@ -674,4 +675,39 @@ public void checkIfEditorGroupsNeedsToBeCreated(Course course) {
...
+     * @return the allowed time span size
+     */
+    public int determineTimeSpanSizeForActiveStudents(Course course, ZonedDateTime endDate, int maximalSize) {
+        var spanTime = maximalSize;
+        if (course.getStartDate() != null) {
+            var startDateIsoWeek = statisticsRepository.getWeekOfDate(course.getStartDate());
+            var endDateIsoWeek = statisticsRepository.getWeekOfDate(endDate);
+            int weeksInYear = Math.toIntExact(IsoFields.WEEK_OF_WEEK_BASED_YEAR.rangeRefinedBy(course.getStartDate()).getMaximum());
+            int amountOfWeeksBetween = (endDateIsoWeek - startDateIsoWeek + weeksInYear) % weeksInYear;
+            spanTime = Math.min(maximalSize, amountOfWeeksBetween + 1);
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/CourseService.java:709]: I think there was something similar some time ago but I don’t remember the details: Why does
```java
course.getStartDate().until(endDate.plusWeeks(1), ChronoUnit.WEEKS)
```
not work in this case?

File: src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts [hunk 2]
```diff
@@ -121,13 +126,18 @@ export class CourseDetailLineChartComponent implements OnChanges {
     }
 
     private createLabels() {
+        this.dataCopy[0].series = [{}];
+        this.absoluteSeries = [{}];
         const prefix = this.translateService.instant('calendar_week');
-        const startDate = dayjs().subtract(this.displayedNumberOfWeeks - 1 + this.displayedNumberOfWeeks * -this.currentPeriod, 'weeks');
-        const endDate = this.currentPeriod !== 0 ? dayjs().subtract(this.displayedNumberOfWeeks * -this.currentPeriod, 'weeks') : dayjs();
+        const endDate = dayjs().subtract(this.currentOffsetToEndDate + this.displayedNumberOfWeeks * -this.currentPeriod, 'weeks');
```
Comment (b-fein) [inline src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts:132]: `this.currentOffsetToEndDate + this.displayedNumberOfWeeks * -this.currentPeriod` appears in the calculation of the `startDate` again (with some other addition in between). Can you move it to a variable with a name that explains what is actually calculated here?

File: src/main/webapp/app/shared/chart/active-students-chart.directive.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import { Directive } from '@angular/core';
+import { Course } from 'app/entities/course.model';
+import dayjs from 'dayjs/esm';
+
+@Directive()
+export class ActiveStudentsChartDirective {
```
Comment (b-fein) [inline src/main/webapp/app/shared/chart/active-students-chart.directive.ts:6]: Does this actually need to be a directive or could it be a regular class?

File: src/test/java/de/tum/in/www1/artemis/util/CourseTestService.java [hunk 4]
```diff
@@ -1620,6 +1620,7 @@ public void testGetExerciseStatsForCourseOverview() throws Exception {
         assertThat(courseDtos).hasSize(1);
         var dto = courseDtos.get(0);
         assertThat(dto.getCourseId()).isEqualTo(instructorsCourse.getId());
+        assertThat(dto.getActiveStudents().size()).as("course was only active for 3 days").isEqualTo(1);
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/util/CourseTestService.java:1623]:
```suggestion
        assertThat(dto.getActiveStudents()).as("course was only active for 3 days").hasSize(1);
```

**Review by b-fein — APPROVED** [review]
Code review: Looks good, though with those date calculation is is likely that small issues like off-by-ones are only noticed during manual testing.
I have left a few comments which are mostly not suggestions but questions to help my understanding.

File: src/test/java/de/tum/in/www1/artemis/util/CourseTestService.java (hunk 4 above)
Comment (willergeorg) [reply src/test/java/de/tum/in/www1/artemis/util/CourseTestService.java]: This seems to be better to me, thanks for the suggestion.

File: src/main/webapp/app/shared/chart/active-students-chart.directive.ts (hunk 3 above)
Comment (willergeorg) [reply src/main/webapp/app/shared/chart/active-students-chart.directive.ts]: Indeed, should be fixed

File: src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts (hunk 2 above)
Comment (willergeorg) [reply src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts]: Done in b9ecc6d

File: src/main/java/de/tum/in/www1/artemis/service/CourseService.java (hunk 1 above)
Comment (willergeorg) [reply src/main/java/de/tum/in/www1/artemis/service/CourseService.java]: I oriented at code that already performed a similar task. I tried your suggestion, which seemed to work for me, and passed the adapted test cases. Changed accordingly in b9ecc6d

**Comment by willergeorg** [comment]
> Code review: Looks good, though with those date calculation is is likely that small issues like off-by-ones are only noticed during manual testing. I have left a few comments which are mostly not suggestions but questions to help my understanding.

Thanks for taking the time and looking at the changes. I updated the code according to your feedback.

**Review by b-fein — APPROVED** [review]
Thanks, code looks good! :+1:

**Review by Dominik-Weinzierl — CHANGES_REQUESTED** [review]
I noticed some issues after step 4 (course with zero participants). Instead of `NaN%` it should be `0%`.

<details>
<summary>Before step 4:</summary>

image
</details>


<details>
<summary>After step 4:</summary>

image
</details>

Edit: This can also be solved in a follow up PR if you want to merge this PR :+1:

**Comment by willergeorg** [comment]
> I noticed some issues after step 4 (course with zero participants). Instead of `NaN%` it should be `0%`.
> 
> Before step 4:
> After step 4:
> Edit: This can also be solved in a follow up PR if you want to merge this PR 👍

Thanks for reporting this issue to me. I refactored the corresponding client code so that NaN should not occur anymore.
I also changed the server test implementation a little bit, because the prior configuration of submission dates led to a flaky server test. If you detect anything else that seems to be unintended, just let me know.

File: src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts [hunk 5]
```diff
@@ -121,13 +127,26 @@ export class CourseDetailLineChartComponent implements OnChanges {
...
+        displayedNumberOfWeeks determines the normal scope of the chart (usually 17 weeks).
+        currentPeriod indicates how many times the observer shifted the scope in the past (by pressing the arrow)
+         */
+        const diffToLastChartWeek = this.currentOffsetToEndDate + this.displayedNumberOfWeeks * -this.currentPeriod;
+        const endDate = dayjs().subtract(diffToLastChartWeek, 'weeks');
+        const remainingWeeksTillStartDate = this.course.startDate ? this.determineDifferenceBetweenIsoWeeks(this.course.startDate, endDate) + 1 : this.displayedNumberOfWeeks;
+        this.currentSpanSize = Math.min(remainingWeeksTillStartDate, this.displayedNumberOfWeeks);
+        // for the start date, we subtract the currently possible span size - 1 from the end date in addition
+        const startDate = dayjs().subtract(diffToLastChartWeek + this.currentSpanSize - 1, 'weeks');
+        this.startDateDisplayed = this.course.startDate ? remainingWeeksTillStartDate <= this.displayedNumberOfWeeks : false;
```
Comment (bassner) [inline src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts:145]:
```suggestion
        this.startDateDisplayed = !!this.course.startDate && remainingWeeksTillStartDate <= this.displayedNumberOfWeeks;
```

**Review by bassner — CHANGES_REQUESTED** [review]
Only a few minor _suggestions_ :)

**Review by Dominik-Weinzierl — APPROVED** [review]
Well done :tada: the problem with `NaN` is solved :+1: (tested locally)

File: src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.html [hunk 6]
```diff
@@ -7,57 +7,70 @@ <h5>{{ 'artemisApp.courseStatistics.activeStudents' | artemisTranslate }}</h5>
...
+            <h1 *ngIf="loading" class="spinner">
+                <fa-icon [icon]="faSpinner" [spin]="true">&nbsp;</fa-icon>
+            </h1>
+            <fa-icon
+                [icon]="faArrowRight"
+                size="2x"
+                class="col-md-1 d-flex justify-content-start align-items-center"
+                role="button"
+                (click)="switchTimeSpan(RIGHT)"
+                [ngClass]="showsCurrentWeek ? 'hide-arrow' : ''"
```
[the remark this answers is not in the document]
Comment (Dominik-Weinzierl) [reply src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.html:61]: You could also use `[class.hide-arrow]="showsCurrentWeek"` - but both ways are good :+1:

**Comment by akesfeden** [comment]
Tested locally, I am not sure if this is related to your changes but when the ratio is 100%, it seems like the line goes over the top of the chart. I would expect the line to be continuous. Can you also reproduce this?

image

**Comment by willergeorg** [comment]
> Tested locally, I am not sure if this is related to your changes but when the ratio is 100%, it seems like the line goes over the top of the chart. I would expect the line to be continuous. Can you also reproduce this?
> 
> image

Thanks for reporting it. This is a known issue unrelated to this PR. The computation of the relative value is based on the current number of enrolled students. If you remove students from the course, the relative value can exceed 100%. As students currently cannot leave a course by themselves and the instructor would need to remove them manually, this is an edge case on production, especially for a large number of students.

**Review by cremertim — DISMISSED** [review]
tested locally according to the testing steps. Works as expected for me

**Review by bassner — DISMISSED** [review]
LGTM 👍

File: src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts [hunk 7]
```diff
@@ -121,13 +127,26 @@ export class CourseDetailLineChartComponent implements OnChanges {
...
+        this.currentSpanSize = Math.min(remainingWeeksTillStartDate, this.displayedNumberOfWeeks);
+        // for the start date, we subtract the currently possible span size - 1 from the end date in addition
+        const startDate = dayjs().subtract(diffToLastChartWeek + this.currentSpanSize - 1, 'weeks');
+        this.startDateDisplayed = !!this.course.startDate && remainingWeeksTillStartDate <= this.displayedNumberOfWeeks;
         let currentWeek;
-        for (let i = 0; i < this.displayedNumberOfWeeks; i++) {
+        for (let i = 0; i < this.currentSpanSize; i++) {
             currentWeek = dayjs()
-                .subtract(this.displayedNumberOfWeeks - 1 + this.displayedNumberOfWeeks * -this.currentPeriod - i, 'weeks')
+                .subtract(this.currentOffsetToEndDate + this.currentSpanSize - 1 + this.displayedNumberOfWeeks * -this.currentPeriod - i, 'weeks')
```
Comment (b-fein) [inline src/main/webapp/app/course/manage/detail/course-detail-line-chart.component.ts:149]:
```suggestion
                .subtract(this.currentOffsetToEndDate + this.currentSpanSize - 1 - this.displayedNumberOfWeeks * this.currentPeriod - i, 'weeks')
```

**Review by akesfeden — APPROVED** [review]
Thank you for the thorough explanation @willergeorg, besides that I haven't seen any issues in my local test.

**Review by b-fein — APPROVED** [review]
Code: Changes since last review look good.

**Review by filip-gregurevic — APPROVED** [review]
Re-approve 👍🏼

**Review by bassner — APPROVED** [review]
Changes since last review make sense 👍

**Review by Dominik-Weinzierl — APPROVED** [review]
Changes since last approve look good :+1:   tested again locally and could not find any new problems. Well done :tada:

**Review by akesfeden — APPROVED** [review]
Code changes look good to me.

**Pull request merged**